### ISSUE TITLE
LMA debug information

### ISSUE THREAD

**Issue by TianBo-Timothy** [OpenANN/OpenANN#178]
I think the output of error value in LMA.cpp:42 
```c++
                  << FloatingPointFormatter(errorValues.mean(), 4);
```
should be
```c++
                  << FloatingPointFormatter(errorValues.squaredNorm()/errorValues.size(), 4);
```

To minimize the problem sum_i (f_i^2(x)) with LMA, one need to provide f_i(x) to the algorithm, instead of f_i^2(x). Thus errorValues holds f_i(x) (i.e. the differences between the neural network output and observed y).

Current code sometimes prints out negative value (running example twospirals).

**Comment by AlexanderFabisch** [comment]
That is because in this example we optimize the cross-entropy and not the sum of squared error. LMA is usually designed for a sum of squared errors. In this case each partial error is already a squared difference to the desired value.

**Comment by TianBo-Timothy** [comment]
I am talking about the example two spirals. It is using MSE instead of CE. see examples/twospirals/TwoSpiralsVisualization.cpp:167
```cpp
    train(*net, "LMA", MSE, stop, true);
```

I need to correct one thing in my post: OpenANN example does not produce negative value - it produce negative value on my test program.

I checked again and noticed something in Net.cpp:526
```cpp
    return meanSquaredError(tempOutput - trainSet->getTarget(n).transpose()) +
        regularizationError;
```

Let's ignore the regularization error. The above line actually returns the squared error. But alglib do not need squared error, instead, it needs the error itself. Suppose there are 4 outputs, there should be 4 terms in LMA solver, each provides an error.

I understand what is done in OpenANN as a trick - maybe you want one single term for one training sample, so the squared error can put terms for all outputs into one solid term. This trick works, but unfortunately, it kills performance - the problem becomes double-quadratic now, and thus need more precision to get a result.

I notice this problem because I developed a demo with two spirals problem with alglib LMA solver. When I tried to migrate my program to OpenANN, I notice that I can get a result with precision set to 1e-4 while using OpenANN's LMA wrapper, I have to set precision to 1e-8 to get a similar result.

My approach to deal with multiple outputs is to consider them as separate terms in LMA solver. Thus we will have n x m functions where n is the number of training samples, and m is the number of outputs. This of cause will increase the size of the problem, but comparing to the double-quadratic problem, the increased size is not a big issue.

Just a reminder: when changing to raw error, the gradient will be changed accordingly.

**Comment by TianBo-Timothy** [comment]
Setting to higher precision means more execution time.

**Comment by AlexanderFabisch** [comment]
OK, sorry, I did not check the example. :)

The problem is that the architectures does not allow to compute an individual error and gradient for each output at the moment. Do you have an idea how to solve that problem?

**Comment by TianBo-Timothy** [comment]
In my implementation, what I did is like this: for LMA, suppose the output size is m, then f_n is for j-th output of i-th sample, where i = n / m, j = n %m. In the forward pass we already have all outputs, so we can keep it and provide them when needed (as long as the parameter does not change).

The code looks like this:

```cpp
double Net::error(unsigned int n)
{
  if(errorFunction == CE)
    return crossEntropyError(n);
  else {
    unsigned int i = n / m; 
    if (i != currentCase) { // currentCase records what we were processing.
       tempInput = trainSet->getInstance(n).transpose();
       forwardPropagate(...);
       currentCase = i;
    }
    unsigned int j = n % m;
    return  tempOutput(j) - trainSet->getTarget(n)(j);
  }
}
```
The regularization terms can be done similarly. 

Since I am not so familiar with the code, I am not if the solution works. Just take it as a hint.

**Comment by TianBo-Timothy** [comment]
I just realised that we can, actually, combine multiple outputs terms into one term. Let f_i(x) be the error for the i-th output of given training example. Denote f(x) as sqrt( sum_i f_i(x) ), so that f(x) can be used as the term for the given training example for LMA.

The derivative of this term is (del f)/(del x) = sum_i (f_i / f) (del f_i)/ (del x). For backpropgation, this means we just need to replace f_i to f_i / f to make it work.

What do you think?

**Comment by AlexanderFabisch** [comment]
I am a little bit confused. :)

At the moment we have f_i = (y_i - t_i) and MSE = 0.5 * \sum_i f_i^2 (without sqrt). Then \partial MSE / \partial f_i = f_i and \partial MSE / \partial weight = \sum_i f_i * (\partial y_i - t_i) / \partial weight.

Would you suggest to replace MSE with \sqrt(MSE) and where would you replace f_i by f_i / f in the gradient calculation?

**Comment by TianBo-Timothy** [comment]
I was talking about the case with multiple outputs. So that MSE = 0.5 \sum_i \sum_j (y_ij - t_ij)^2, where y_ij is the j-th output for i-th training example.

To utilise LMA, we need to set f_k = y_ij - t_ij, where k = i * n + j, n is the number of outputs. This is inconvenient, as we have to calculate n different derivatives for each training example, i.e. we need to calculate \partial f_(i*n) / \partial weight, \partial f_(i*n + 1) / \partial weight, ... \partial f_(i*n + n-1) / \partial weight. If I am correct, this needs multiple backpropgations.

The last comment is to suggest combine the terms for each training example. That is to let f_i = sqrt(\sum_j f_ij^2), where f_ij = y_ij - t_ij. In this way, MSE will not change as MSE = 0.5 \sum_i f_j^2 = 0.5 \sum_i \sum_j (y_ij - t_ij)^2. We are now solving the problem MSE = 0.5 \sum_i f_i^2, which perfectly fits to LMA.

The form of f_i (square root of sum of squares) has a nice derivative property that
\partial f_i / \partial weight = sum_j (\partial f_i / \partial f_ij) (\partial f_ij / \partial weight) = sum_j (f_ij / f_i) (\partial f_ij / \partial weight). So that we can get gradients with one backpropgation by replacing f_ij by f_ij / f_i.

**Comment by AlexanderFabisch** [comment]
The optimization immediately stops when I put sqrt around the MSE in Net.cpp:526. Probably some weird bug. I don't have time to look into this at the moment. The other change I would do is something like this after line 596:

    if(errorFunction == CE)
        tempError /= value;

Right?

That however, would break the implementation for all other types of optimizers. Currently I would say it's not worth the effort for me.
